### DISCUSSION TITLE
improve npm performance vs pnpm

### DISCUSSION THREAD

**Discussion by mariusa** [npm/feedback#341] [CLI] [answered]
`pnpm` is more than twice as fast on clean installs: https://pnpm.io/benchmarks

Could `npm` performance be closer to `pnpm`'s ?

Benefits of pnpm:

* It’s compatible with npm but also offers significantly better disk space usage and speed

* pnpm installs all packages on a single location and then uses symlinks to reference them. It introduces a completely new concept called a content-addressable storage system that enables pnpm to detect the difference between files. In turn, it doesn’t duplicate unchanged files in two different versions of a package

* pnpm has a strict access control mechanism, meaning that a package can only access the dependencies that are defined in its package.json file

**Comment by ljharb** [comment]
pnpm and yarn pnp's symlink approach breaks various parts of the ecosystem in subtle ways; it's not a trivial change, and would likely break a lot of valid use cases (including, explicitly, the ability for a package to require something not declared in its own dependencies list)

**Reply by mariusa** [reply]
I see. I never used it, by it keeps popping up on various projects, more often. Would be great to get closer in terms of performance.

**Reply by ljharb** [reply]
Performance is great, as long as it doesn't sacrifice correctness :-)

**Reply by Brassrat** [reply]
I'm not sure if pnpm breaks _valid_ use cases, it clearly breaks situations where an app requires something not explicitly in its package.json but happens to be required by some package which may be only implicitly included. This causes implicit dependencies on versions which may change willy-nilly based on changes to these other/unknown packages. Very hard to track down.
I can't see a use-case for being able to require something you didn't specify in your package.json unless you are explicitly acquiring the required artifact in some other manner. I don't think depending on some artifact to include some other artifact is valid design. Sort of runs up against the Law of Demeter

**Reply by ljharb** [reply]
That is indeed a valid use case. Any time a package wants to require something a user may have installed, this is something they'd want to do. Only _very_ recently did peerDependenciesMeta land which allows peer deps to be optional (and not installed by default), so there will be a number of packages that haven't used this to explicitly list a dep, and a lot that _can't_ because `peerDependencies` are required, and there's still many npm versions out there that don't understand `peerDependenciesMeta`.

**Reply by Brassrat** [reply]
As i wrote, depending on an artifact that is expected to be acquired by some other mechanism can make sense. What doesn't make sense, to me, is depending on an artifact that some UNKNOWN artifact depends on. So someone comes on to a project and decides they want to use artifact X, so they require('X') but forget to add X to the project's dependencies, but some other artifact Y depends on Z which depends on X. All appears well until Y or Z changes, possibly to not depend on X or upgrade to a new version. All of a sudden the app no longer runs because X is missing? or its behavior has changed. Not fun tracking this down.

**Reply by ljharb** [reply]
Sure, I understand that there are many use cases where the omission is a bug! However, there are nonzero use cases where it's intentional, which simply makes pnpm/yarn's decision here one that breaks valid use cases. They have clearly decided it's worth the cost/tradeoff, but I hope npm never considers that cost acceptable.

**Reply by Brassrat** [reply]
I guess i don't understand the good use cases that pnpm causes breakage. Npm won't conjure up artifacts that are not in package.json so if there are needed artifacts that are not in the package.json how do they become available? Either they are put in the 'node search path' by another part of the 'build' process (which is fine, i.e., the dependency is explicit) or obtained dynamically (also fine) or you are depending on npm's decision to essentially make sub-artifacts appear to be top-level or direct artifacts. Since this is a non-deterministic process (if the artifact in question is a dependency of more than one artifact, they it depends upon which 'parent' artifact is processed first which may change) I feel that depending on this behavior is not good practice (however, convenient.)
Obviously, if the 'parent' artifact is one's own then we are ok to have some knowledge of what it provides, but then by the Law of Demeter one should require the 'parent' and then ask it (the parent) for the desired artifact.

**Reply by ljharb** [reply]
It is non-determinstic, but that's not inherently invalid or "bad".

**Comment by MylesBorins** [comment]
@mariusa the npm team is always interested in finding ways to improve performance of npm, but as mentioned by @ljharb we are limited in ways of doing this to ensure that we don't break existing installations or platforms.

I'll let the @npm/cli team chime in with some details about how we plan to measure speed and different ways we've thought about improving the overall experience

**Comment by darcyclarke** [comment, marked as answer]
@mariusa to quickly start out, & as @ljharb noted, the way each package manager resolves dependencies is different resulting in different checks/evaluations, tress &, ultimately, packages on disk. With that in mind, performance can be a tough thing to track accurately & compare since these tools all have different approaches & understandings of correctness (ex. `pnpm` & `yarn` do not audit packages or install peer dependencies by default, so you have to provide flags like `--no-audit` & `--legacy-peer-deps` to get a semblance of a fair result); Moreso, there are numerous factors that play into performance benchmarking which make it even more of a ficle metric (including, but not limited to: environment/system-specific specs/config effecting network throughput, disk speed etc. & a significant number of project-specific details that come into play based on the approaches taken by each pm).

In terms of accuracy for performance, I would try running our [benchmark suite](https://github.com/npm/benchmarks) to ensure there isn't a heavy bias toward only a single data point or method of evaluation (ie. only referencing `pnpm`'s data); And I'd also encourage you to take one of your own project's existing `package.json` & try testing that as a fixture with that tool.

Topping up on @MylesBorins, `npm` is constantly considering new _strategies_ that will help make developer's lives easier; We have indeed been considering how we could introduce a similar single _store_, _linking_ & _strictness_ _model_ to `pnpm`'s without breaking our existing reification _strategy_; There will be [RFCs](https://github.com/npm/rfcs) created in the **very near future** which outline how we believe we could move forward with this.

> Note: a good trend to see here as well is that we constantly are tracking our progress of performance against ourselves; Here's a view of the diff between `v6.14.12` & `v7.7.6`: https://github.com/npm/benchmarks/issues/28, And, fwiw, we know there's a lot more we can do to improve those numbers without resorting to radically changing our approach.

**Reply by mariusa** [reply]
Thanks for the details and considering new strategies! Hope more developers will settle on `npm` for their projects.

**Comment by mariusa** [comment]
As an update, I keep seeing projects switching from npm to yarn or pnpm, eg https://github.com/sequelize/sequelize/pull/14036#issuecomment-1040038861

with motivations such as `yarn is quicker overall and the lockfile is nicer.`

Would love to have a single standard package manager, instead of having efforts & learning curve of new node developers fragmented across many choices. Correctness or not, reality is npm is getting more and more excluded out :(
